File: DSA/src/Questions/BinaryTree/RightView.java
package Questions.BinaryTree;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class RightView {

    public List<Integer> rightSideView(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()) {
            int size = q.size();
            for (int elem = 0; elem < size; elem++) {
                TreeNode curr = q.poll();
                assert curr != null;
                if (elem == size - 1) {
                    result.add(curr.val);
                }

                if (curr.left != null) {
                    q.add(curr.left);
                }
                ;
                if (curr.right != null) {
                    q.add(curr.right);
                }
                ;

            }

        }
        return result;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        TreeNode right = new TreeNode(2);
        TreeNode rightLeft = new TreeNode(3);
        root.right = right;
        right.left = rightLeft;
        RightView rightView = new RightView();
        List<Integer> result = rightView.rightSideView(root);
        System.out.println("Right View: " + result);
    }
}

File: DSA/Python/Questions/BinaryTree/RightView.py
"""
# Question: Binary Tree Right Side View
# Link: https://leetcode.com/problems/binary-tree-right-side-view/

# Time Complexity: O(N)
# Space Complexity: O(H)

# Algorithm:
# 1. Level order traversal
# 2. Track rightmost nodes
# 3. DFS with level tracking
"""

from collections import deque
from typing import List


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class RightView:
    def right_side_view(self, root: TreeNode) -> List[int]:
        if not root:
            return []

        result = []
        max_level = [-1]

        def dfs(node: TreeNode, level: int) -> None:
            if not node:
                return

            if level > max_level[0]:
                result.append(node.val)
                max_level[0] = level

            dfs(node.right, level + 1)
            dfs(node.left, level + 1)

        dfs(root, 0)
        return result


def main():
    solution = RightView()
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    print(solution.right_side_view(root))


if __name__ == "__main__":
    main()
